src/query/storage/m3/dynamic_cluster_test.go

```diff
@@ -0,0 +1,416 @@
+		{nsID: fooNsID, nsOpts: fooOpts},
+		{nsID: barNsID, nsOpts: barOpts},
+	})
+
+	mapCh2 := make(nsMapCh, 10)
+	mapCh2 <- nsMap
+	mockInitializer2 := newMockNsInitializer(t, ctrl, mapCh2)
+
+	cfg := DynamicClusterNamespaceConfiguration{
+		session:       mockSession,
+		nsInitializer: mockInitializer,
+	}
+	cfg2 := DynamicClusterNamespaceConfiguration{
+		session:       mockSession2,
+		nsInitializer: mockInitializer2,
+	}
+
+	opts := NewDynamicClusterOptions().
+		SetDynamicClusterNamespaceConfiguration([]DynamicClusterNamespaceConfiguration{cfg, cfg2}).
+		SetInstrumentOptions(instrument.NewOptions())
+
+	clusters, err := newDynamicClusters(opts)
+	require.NoError(t, err)
+
+	defer cleanup(clusters)
+
+	requireClusterNamespace(t, clusters, defaultTestNs2ID, ClusterNamespaceOptions{
+		attributes: storagemetadata.Attributes{
+			MetricsType: storagemetadata.AggregatedMetricsType,
+			Retention:   48 * time.Hour,
+			Resolution:  1 * time.Minute,
+		},
+		downsample: &ClusterNamespaceDownsampleOptions{All: true},
+	})
+
+	requireClusterNamespace(t, clusters, fooNsID, ClusterNamespaceOptions{
+		attributes: storagemetadata.Attributes{
+			MetricsType: storagemetadata.AggregatedMetricsType,
+			Retention:   48 * time.Hour,
+			Resolution:  2 * time.Minute,
+		},
+		downsample: &ClusterNamespaceDownsampleOptions{All: true},
+	})
+
+	requireClusterNamespace(t, clusters, barNsID, ClusterNamespaceOptions{
+		attributes: storagemetadata.Attributes{
+			MetricsType: storagemetadata.AggregatedMetricsType,
+			Retention:   48 * time.Hour,
+			Resolution:  5 * time.Minute,
+		},
+		downsample: &ClusterNamespaceDownsampleOptions{All: false},
+	})
+
+	requireClusterNamespace(t, clusters, defaultTestNs1ID, ClusterNamespaceOptions{
+		attributes: storagemetadata.Attributes{
+			MetricsType: storagemetadata.UnaggregatedMetricsType,
+			Retention:   48 * time.Hour,
+		}})
+
+	requireClusterNamespaceIDs(t, clusters, []ident.ID{defaultTestNs1ID, defaultTestNs2ID,
+		fooNsID, barNsID})
+}
+
+func TestDynamicClustersInitFailures(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	mockSession := client.NewMockSession(ctrl)
+
+	updateCh := make(chan struct{}, 10)
+	reg := namespace.NewMockRegistry(ctrl)
+	reg.EXPECT().Watch().Return(nil, errors.New("failed to init")).AnyTimes()
+
+	cfg := DynamicClusterNamespaceConfiguration{
+		session: mockSession,
+		nsInitializer: &mockNsInitializer{
+			registry: reg,
+			updateCh: updateCh,
+		},
+	}
+
+	opts := NewDynamicClusterOptions().
+		SetDynamicClusterNamespaceConfiguration([]DynamicClusterNamespaceConfiguration{cfg}).
+		SetInstrumentOptions(instrument.NewOptions())
+
+	_, err := newDynamicClusters(opts)
+	require.Error(t, err)
+}
+
+func cleanup(clusters Clusters) {
+	// Give test a chance to schedule pending goroutines (i.e. goroutine watching for updates)
+	// If this is not done, test may complete and Close() will be called before startWatch goroutine
```

Comment:
Can this be hardened by adding a `updateChan chan struct` to each test, so it won't call cleanup until a watch has triggered?

Reply:
I think the only way to get this to work requires modifying the non-test code but let me think on it a bit more.